tools/lsp/common.rs

```diff
@@ -25,32 +28,142 @@ pub type UrlVersion = Option<i32>;
+        Self(type_loader)
+    }
+
+    pub fn resolve_import_path(
+        &self,
+        import_token: Option<&i_slint_compiler::parser::NodeOrToken>,
+        maybe_relative_path_or_url: &str,
+    ) -> Option<(PathBuf, Option<&'static [u8]>)> {
+        self.0.resolve_import_path(import_token, maybe_relative_path_or_url)
+    }
+
+    pub fn document_version(&self, target_uri: &Url) -> SourceFileVersion {
+        self.0
             .get_document(&uri_to_file(target_uri).unwrap_or_default())
             .and_then(|doc| doc.node.as_ref()?.source_file.version())
     }
+
+    pub fn get_document(&self, url: &Url) -> Option<&Document> {
+        let path = uri_to_file(url)?;
+        self.0.get_document(&path)
+    }
+
+    pub fn get_document_by_path(&self, path: &Path) -> Option<&Document> {
+        self.0.get_document(&path)
+    }
+
+    pub fn get_document_for_source_file(&self, source_file: &SourceFile) -> Option<&Document> {
+        self.0.get_document(source_file.path())
+    }
+
+    pub fn all_url_documents(&self) -> impl Iterator<Item = (Url, &Document)> + '_ {
+        self.0.all_file_documents().filter_map(|(p, d)| Some((file_to_uri(p)?, d)))
+    }
+
+    pub fn all_urls(&self) -> impl Iterator<Item = Url> + '_ {
+        self.0.all_files().filter_map(|p| file_to_uri(p))
+    }
+
+    pub fn global_type_registry(&self) -> std::cell::Ref<TypeRegister> {
+        self.0.global_type_registry.borrow()
+    }
+
+    pub async fn reconfigure(
+        &mut self,
+        style: Option<String>,
+        include_paths: Option<Vec<PathBuf>>,
+        library_paths: Option<HashMap<String, PathBuf>>,
+    ) -> Result<CompilerConfiguration> {
+        if style.is_none() && include_paths.is_none() && library_paths.is_none() {
+            return Ok(self.0.compiler_config.clone());
+        }
+
+        if let Some(s) = style {
+            if s.is_empty() {
+                self.0.compiler_config.style = None;
+            } else {
+                self.0.compiler_config.style = Some(s);
+            }
+        }
+
+        if let Some(ip) = include_paths {
+            self.0.compiler_config.include_paths = ip;
+        }
+
+        if let Some(lp) = library_paths {
+            self.0.compiler_config.library_paths = lp;
+        }
+
+        self.preload_builtins().await;
+
+        Ok(self.0.compiler_config.clone())
+    }
+
+    pub async fn preload_builtins(&mut self) {
+        // Always load the widgets so we can auto-complete them
+        let mut diag = BuildDiagnostics::default();
+        self.0.import_component("std-widgets.slint", "StyleMetrics", &mut diag).await;
+        assert!(!diag.has_error());
+    }
+
+    pub async fn load_url(
+        &mut self,
+        url: &Url,
+        version: SourceFileVersion,
+        content: String,
+        diag: &mut BuildDiagnostics,
+    ) -> Result<()> {
+        let path =
+            uri_to_file(url).ok_or::<Error>(String::from("Failed to convert path").into())?;
```

Comment:
seems overly complicated.  
Why don't `uri_to_file(url).ok_or("Failed to convert path")?;` work?  The `?` already have an into builtin so you shouldn't need the into.  and i think we can convert from &str directly.  Also use .to_string or .to_owned instead of String::from